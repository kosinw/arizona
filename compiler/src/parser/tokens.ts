import * as moo from 'moo';
import { SyntaxNativeType } from './types';

export const keywords = [
  'break',
  'if',
  'else',
  'use',
  'from',
  'export',
  'return',
  'const',
  'let',
  'for',
  'continue',
  'while',
  'fn',
  'type',
  'struct',
];

export const punctuators = [
  '+',
  '++',
  '-',
  '--',
  '>>',
  '<<',
  '>>=',
  '<<=',
  '=',
  '==',
  '+=',
  '-=',
  '=>',
  '<=',
  '>=',
  '!=',
  '%',
  '%=',
  '*',
  '*=',
  '/',
  '/=',
  '^',
  '^=',
  '&',
  '&=',
  '~',
  '|',
  '|=',
  '!',
  ':',
  '(',
  ')',
  '.',
  '{',
  '}',
  ',',
  '[',
  ']',
  ';',
  '>',
  '<',
  '?',
  '||',
  '&&',
  '{',
  '}',
  '->',
];

export const types = Object.values(SyntaxNativeType);

export const tokens: moo.Rules = {
  whitespace: { match: /\s+/, lineBreaks: true },
  comment: [
    { match: /\/\/.*?$/ },
    { match: /\/\*[^]*?\*\//, lineBreaks: true },
  ],
  float: { match: /[0-9]+\.[0-9]*/ },
  integer: [
    { match: /0[xX][0-9a-fA-F_]+/ },
    { match: /0[oO][0-7_]+/ },
    { match: /0[bB][01_]+/ },
    { match: /(?:[1-9][0-9_]*|0)/ },
  ],
  char: [
    { match: /'(?:\\['\\bnfrtv0]|[^'\\\n])'/, value: (x) => x.slice(1, -1) },
  ],
  string: [
    { match: /"(?:\\["\\rn]|[^"\\\n])*?"/, value: (x) => x.slice(1, -1) },
    { match: /'(?:\\['\\bfnrtv0]|[^'\\\n])*?'/, value: (x) => x.slice(1, -1) },
  ],
  bool: [{ match: /true|false/ }],
  identifier: {
    match: /[A-Za-z_][A-Za-z0-9_]*/,
    type: moo.keywords({ keyword: keywords, type: types }),
  },
  punctuator: punctuators,
};
