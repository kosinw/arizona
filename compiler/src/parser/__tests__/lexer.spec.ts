import { describe, it, expect } from '@jest/globals';
import dedent from 'dedent';
import { tokens } from '../tokens';
import * as moo from 'moo';

describe('lexer tests', () => {
  it('lexes some tokens', () => {
    const src = `
    // comment signle line
    /* hello
    */ world
    const let a = 32; // ''
    lf\ncr\rcrlf\r\n
    +-=_|
    `;

    const lexer = moo.compile(tokens);
    lexer.reset(src);

    expect(Array.from(lexer)).toMatchSnapshot();
  });

  it('lexes more tokens', () => {
    const src = dedent`
        fn factorial(n: i32) -> i32 {
            let j = 15;

            if (n == 1 || n == 0) {
                return 1;
            }

            return n * factorial(n-1)
        }
    `;

    const lexer = moo.compile(tokens);
    lexer.reset(src);

    expect(Array.from(lexer)).toMatchSnapshot();
  });
});
