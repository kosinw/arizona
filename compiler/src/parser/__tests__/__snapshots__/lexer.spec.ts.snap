// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer tests lexes more tokens 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "fn",
    "toString": [Function],
    "type": "keyword",
    "value": "fn",
  },
  Object {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "factorial",
    "toString": [Function],
    "type": "identifier",
    "value": "factorial",
  },
  Object {
    "col": 13,
    "line": 1,
    "lineBreaks": 0,
    "offset": 12,
    "text": "(",
    "toString": [Function],
    "type": "punctuator",
    "value": "(",
  },
  Object {
    "col": 14,
    "line": 1,
    "lineBreaks": 0,
    "offset": 13,
    "text": "n",
    "toString": [Function],
    "type": "identifier",
    "value": "n",
  },
  Object {
    "col": 15,
    "line": 1,
    "lineBreaks": 0,
    "offset": 14,
    "text": ":",
    "toString": [Function],
    "type": "punctuator",
    "value": ":",
  },
  Object {
    "col": 16,
    "line": 1,
    "lineBreaks": 0,
    "offset": 15,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 17,
    "line": 1,
    "lineBreaks": 0,
    "offset": 16,
    "text": "i32",
    "toString": [Function],
    "type": "type",
    "value": "i32",
  },
  Object {
    "col": 20,
    "line": 1,
    "lineBreaks": 0,
    "offset": 19,
    "text": ")",
    "toString": [Function],
    "type": "punctuator",
    "value": ")",
  },
  Object {
    "col": 21,
    "line": 1,
    "lineBreaks": 0,
    "offset": 20,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 22,
    "line": 1,
    "lineBreaks": 0,
    "offset": 21,
    "text": "->",
    "toString": [Function],
    "type": "punctuator",
    "value": "->",
  },
  Object {
    "col": 24,
    "line": 1,
    "lineBreaks": 0,
    "offset": 23,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 25,
    "line": 1,
    "lineBreaks": 0,
    "offset": 24,
    "text": "i32",
    "toString": [Function],
    "type": "type",
    "value": "i32",
  },
  Object {
    "col": 28,
    "line": 1,
    "lineBreaks": 0,
    "offset": 27,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 29,
    "line": 1,
    "lineBreaks": 0,
    "offset": 28,
    "text": "{",
    "toString": [Function],
    "type": "punctuator",
    "value": "{",
  },
  Object {
    "col": 30,
    "line": 1,
    "lineBreaks": 1,
    "offset": 29,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
  Object {
    "col": 5,
    "line": 2,
    "lineBreaks": 0,
    "offset": 34,
    "text": "let",
    "toString": [Function],
    "type": "keyword",
    "value": "let",
  },
  Object {
    "col": 8,
    "line": 2,
    "lineBreaks": 0,
    "offset": 37,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 9,
    "line": 2,
    "lineBreaks": 0,
    "offset": 38,
    "text": "j",
    "toString": [Function],
    "type": "identifier",
    "value": "j",
  },
  Object {
    "col": 10,
    "line": 2,
    "lineBreaks": 0,
    "offset": 39,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 11,
    "line": 2,
    "lineBreaks": 0,
    "offset": 40,
    "text": "=",
    "toString": [Function],
    "type": "punctuator",
    "value": "=",
  },
  Object {
    "col": 12,
    "line": 2,
    "lineBreaks": 0,
    "offset": 41,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 13,
    "line": 2,
    "lineBreaks": 0,
    "offset": 42,
    "text": "15",
    "toString": [Function],
    "type": "integer",
    "value": "15",
  },
  Object {
    "col": 15,
    "line": 2,
    "lineBreaks": 0,
    "offset": 44,
    "text": ";",
    "toString": [Function],
    "type": "punctuator",
    "value": ";",
  },
  Object {
    "col": 16,
    "line": 2,
    "lineBreaks": 2,
    "offset": 45,
    "text": "

    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "

    ",
  },
  Object {
    "col": 5,
    "line": 4,
    "lineBreaks": 0,
    "offset": 51,
    "text": "if",
    "toString": [Function],
    "type": "keyword",
    "value": "if",
  },
  Object {
    "col": 7,
    "line": 4,
    "lineBreaks": 0,
    "offset": 53,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 8,
    "line": 4,
    "lineBreaks": 0,
    "offset": 54,
    "text": "(",
    "toString": [Function],
    "type": "punctuator",
    "value": "(",
  },
  Object {
    "col": 9,
    "line": 4,
    "lineBreaks": 0,
    "offset": 55,
    "text": "n",
    "toString": [Function],
    "type": "identifier",
    "value": "n",
  },
  Object {
    "col": 10,
    "line": 4,
    "lineBreaks": 0,
    "offset": 56,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 11,
    "line": 4,
    "lineBreaks": 0,
    "offset": 57,
    "text": "==",
    "toString": [Function],
    "type": "punctuator",
    "value": "==",
  },
  Object {
    "col": 13,
    "line": 4,
    "lineBreaks": 0,
    "offset": 59,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 14,
    "line": 4,
    "lineBreaks": 0,
    "offset": 60,
    "text": "1",
    "toString": [Function],
    "type": "integer",
    "value": "1",
  },
  Object {
    "col": 15,
    "line": 4,
    "lineBreaks": 0,
    "offset": 61,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 16,
    "line": 4,
    "lineBreaks": 0,
    "offset": 62,
    "text": "||",
    "toString": [Function],
    "type": "punctuator",
    "value": "||",
  },
  Object {
    "col": 18,
    "line": 4,
    "lineBreaks": 0,
    "offset": 64,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 19,
    "line": 4,
    "lineBreaks": 0,
    "offset": 65,
    "text": "n",
    "toString": [Function],
    "type": "identifier",
    "value": "n",
  },
  Object {
    "col": 20,
    "line": 4,
    "lineBreaks": 0,
    "offset": 66,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 21,
    "line": 4,
    "lineBreaks": 0,
    "offset": 67,
    "text": "==",
    "toString": [Function],
    "type": "punctuator",
    "value": "==",
  },
  Object {
    "col": 23,
    "line": 4,
    "lineBreaks": 0,
    "offset": 69,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 24,
    "line": 4,
    "lineBreaks": 0,
    "offset": 70,
    "text": "0",
    "toString": [Function],
    "type": "integer",
    "value": "0",
  },
  Object {
    "col": 25,
    "line": 4,
    "lineBreaks": 0,
    "offset": 71,
    "text": ")",
    "toString": [Function],
    "type": "punctuator",
    "value": ")",
  },
  Object {
    "col": 26,
    "line": 4,
    "lineBreaks": 0,
    "offset": 72,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 27,
    "line": 4,
    "lineBreaks": 0,
    "offset": 73,
    "text": "{",
    "toString": [Function],
    "type": "punctuator",
    "value": "{",
  },
  Object {
    "col": 28,
    "line": 4,
    "lineBreaks": 1,
    "offset": 74,
    "text": "
        ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
        ",
  },
  Object {
    "col": 9,
    "line": 5,
    "lineBreaks": 0,
    "offset": 83,
    "text": "return",
    "toString": [Function],
    "type": "keyword",
    "value": "return",
  },
  Object {
    "col": 15,
    "line": 5,
    "lineBreaks": 0,
    "offset": 89,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 16,
    "line": 5,
    "lineBreaks": 0,
    "offset": 90,
    "text": "1",
    "toString": [Function],
    "type": "integer",
    "value": "1",
  },
  Object {
    "col": 17,
    "line": 5,
    "lineBreaks": 0,
    "offset": 91,
    "text": ";",
    "toString": [Function],
    "type": "punctuator",
    "value": ";",
  },
  Object {
    "col": 18,
    "line": 5,
    "lineBreaks": 1,
    "offset": 92,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
  Object {
    "col": 5,
    "line": 6,
    "lineBreaks": 0,
    "offset": 97,
    "text": "}",
    "toString": [Function],
    "type": "punctuator",
    "value": "}",
  },
  Object {
    "col": 6,
    "line": 6,
    "lineBreaks": 2,
    "offset": 98,
    "text": "

    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "

    ",
  },
  Object {
    "col": 5,
    "line": 8,
    "lineBreaks": 0,
    "offset": 104,
    "text": "return",
    "toString": [Function],
    "type": "keyword",
    "value": "return",
  },
  Object {
    "col": 11,
    "line": 8,
    "lineBreaks": 0,
    "offset": 110,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 12,
    "line": 8,
    "lineBreaks": 0,
    "offset": 111,
    "text": "n",
    "toString": [Function],
    "type": "identifier",
    "value": "n",
  },
  Object {
    "col": 13,
    "line": 8,
    "lineBreaks": 0,
    "offset": 112,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 14,
    "line": 8,
    "lineBreaks": 0,
    "offset": 113,
    "text": "*",
    "toString": [Function],
    "type": "punctuator",
    "value": "*",
  },
  Object {
    "col": 15,
    "line": 8,
    "lineBreaks": 0,
    "offset": 114,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 16,
    "line": 8,
    "lineBreaks": 0,
    "offset": 115,
    "text": "factorial",
    "toString": [Function],
    "type": "identifier",
    "value": "factorial",
  },
  Object {
    "col": 25,
    "line": 8,
    "lineBreaks": 0,
    "offset": 124,
    "text": "(",
    "toString": [Function],
    "type": "punctuator",
    "value": "(",
  },
  Object {
    "col": 26,
    "line": 8,
    "lineBreaks": 0,
    "offset": 125,
    "text": "n",
    "toString": [Function],
    "type": "identifier",
    "value": "n",
  },
  Object {
    "col": 27,
    "line": 8,
    "lineBreaks": 0,
    "offset": 126,
    "text": "-1",
    "toString": [Function],
    "type": "integer",
    "value": "-1",
  },
  Object {
    "col": 29,
    "line": 8,
    "lineBreaks": 0,
    "offset": 128,
    "text": ")",
    "toString": [Function],
    "type": "punctuator",
    "value": ")",
  },
  Object {
    "col": 30,
    "line": 8,
    "lineBreaks": 1,
    "offset": 129,
    "text": "
",
    "toString": [Function],
    "type": "whitespace",
    "value": "
",
  },
  Object {
    "col": 1,
    "line": 9,
    "lineBreaks": 0,
    "offset": 130,
    "text": "}",
    "toString": [Function],
    "type": "punctuator",
    "value": "}",
  },
]
`;

exports[`lexer tests lexes some tokens 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 1,
    "offset": 0,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
  Object {
    "col": 5,
    "line": 2,
    "lineBreaks": 0,
    "offset": 5,
    "text": "// comment signle line",
    "toString": [Function],
    "type": "comment",
    "value": "// comment signle line",
  },
  Object {
    "col": 27,
    "line": 2,
    "lineBreaks": 1,
    "offset": 27,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
  Object {
    "col": 5,
    "line": 3,
    "lineBreaks": 1,
    "offset": 32,
    "text": "/* hello
    */",
    "toString": [Function],
    "type": "comment",
    "value": "/* hello
    */",
  },
  Object {
    "col": 7,
    "line": 4,
    "lineBreaks": 0,
    "offset": 47,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 8,
    "line": 4,
    "lineBreaks": 0,
    "offset": 48,
    "text": "world",
    "toString": [Function],
    "type": "identifier",
    "value": "world",
  },
  Object {
    "col": 13,
    "line": 4,
    "lineBreaks": 1,
    "offset": 53,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
  Object {
    "col": 5,
    "line": 5,
    "lineBreaks": 0,
    "offset": 58,
    "text": "const",
    "toString": [Function],
    "type": "keyword",
    "value": "const",
  },
  Object {
    "col": 10,
    "line": 5,
    "lineBreaks": 0,
    "offset": 63,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 11,
    "line": 5,
    "lineBreaks": 0,
    "offset": 64,
    "text": "let",
    "toString": [Function],
    "type": "keyword",
    "value": "let",
  },
  Object {
    "col": 14,
    "line": 5,
    "lineBreaks": 0,
    "offset": 67,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 15,
    "line": 5,
    "lineBreaks": 0,
    "offset": 68,
    "text": "a",
    "toString": [Function],
    "type": "identifier",
    "value": "a",
  },
  Object {
    "col": 16,
    "line": 5,
    "lineBreaks": 0,
    "offset": 69,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 17,
    "line": 5,
    "lineBreaks": 0,
    "offset": 70,
    "text": "=",
    "toString": [Function],
    "type": "punctuator",
    "value": "=",
  },
  Object {
    "col": 18,
    "line": 5,
    "lineBreaks": 0,
    "offset": 71,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 19,
    "line": 5,
    "lineBreaks": 0,
    "offset": 72,
    "text": "32",
    "toString": [Function],
    "type": "integer",
    "value": "32",
  },
  Object {
    "col": 21,
    "line": 5,
    "lineBreaks": 0,
    "offset": 74,
    "text": ";",
    "toString": [Function],
    "type": "punctuator",
    "value": ";",
  },
  Object {
    "col": 22,
    "line": 5,
    "lineBreaks": 0,
    "offset": 75,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "col": 23,
    "line": 5,
    "lineBreaks": 0,
    "offset": 76,
    "text": "// ''",
    "toString": [Function],
    "type": "comment",
    "value": "// ''",
  },
  Object {
    "col": 28,
    "line": 5,
    "lineBreaks": 1,
    "offset": 81,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
  Object {
    "col": 5,
    "line": 6,
    "lineBreaks": 0,
    "offset": 86,
    "text": "lf",
    "toString": [Function],
    "type": "identifier",
    "value": "lf",
  },
  Object {
    "col": 7,
    "line": 6,
    "lineBreaks": 1,
    "offset": 88,
    "text": "
",
    "toString": [Function],
    "type": "whitespace",
    "value": "
",
  },
  Object {
    "col": 1,
    "line": 7,
    "lineBreaks": 0,
    "offset": 89,
    "text": "cr",
    "toString": [Function],
    "type": "identifier",
    "value": "cr",
  },
  Object {
    "col": 3,
    "line": 7,
    "lineBreaks": 0,
    "offset": 91,
    "text": "
",
    "toString": [Function],
    "type": "whitespace",
    "value": "
",
  },
  Object {
    "col": 4,
    "line": 7,
    "lineBreaks": 0,
    "offset": 92,
    "text": "crlf",
    "toString": [Function],
    "type": "identifier",
    "value": "crlf",
  },
  Object {
    "col": 8,
    "line": 7,
    "lineBreaks": 2,
    "offset": 96,
    "text": "

    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "

    ",
  },
  Object {
    "col": 5,
    "line": 9,
    "lineBreaks": 0,
    "offset": 103,
    "text": "+",
    "toString": [Function],
    "type": "punctuator",
    "value": "+",
  },
  Object {
    "col": 6,
    "line": 9,
    "lineBreaks": 0,
    "offset": 104,
    "text": "-=",
    "toString": [Function],
    "type": "punctuator",
    "value": "-=",
  },
  Object {
    "col": 8,
    "line": 9,
    "lineBreaks": 0,
    "offset": 106,
    "text": "_",
    "toString": [Function],
    "type": "identifier",
    "value": "_",
  },
  Object {
    "col": 9,
    "line": 9,
    "lineBreaks": 0,
    "offset": 107,
    "text": "|",
    "toString": [Function],
    "type": "punctuator",
    "value": "|",
  },
  Object {
    "col": 10,
    "line": 9,
    "lineBreaks": 1,
    "offset": 108,
    "text": "
    ",
    "toString": [Function],
    "type": "whitespace",
    "value": "
    ",
  },
]
`;
